### ISSUE TITLE
Support Lifecycle Methods in Test Classes

### ISSUE THREAD

**Issue by jlink** [jlink/jqwik#92]
## Testing Problem

With https://github.com/jlink/jqwik/issues/61 done there is the possibility to add 
hooks to container classes or property methods through the `@AddLifecycleHook`
annotation. What's missing is a simple way to influence the lifecycle of just the
properties at hand.

## Suggested Solution

Add annotations that can used on methods in a test classe:

- Container lifecycle
  - `@BeforeContainer` on static methods
  - `@AfterContainer` on static methods

- Around property lifecycle
  - `@BeforeProperty` methods
  - `@AfterProperty` methods

- Around try lifecycle
  - `@BeforeTry` methods
  - `@AfterTry` methods

Parameter resolution should also work for these methods.

## Discussion

In https://github.com/jlink/jqwik/issues/61 there was also a discussion if inner class
based lifecycles wouldn't be better. For the time being the "classic" approach with
annotated methods was considered to be better.

**Comment by gtrefs** [comment]
I am not sure, if this is the right issue for this comment. In #61 @emmmile was  asking about something similar to Jupiters extensions. How could you support this with lifecycle methods? Why am I asking? In Spring the [TestContextManager](https://github.com/spring-projects/spring-framework/blob/master/spring-test/src/main/java/org/springframework/test/context/TestContextManager.java) is the main eintry point to integration testing with Spring. AFAIK it creates an application context for your tests. It offers some methods to control the lifecycle of this context. If I would like to use the framework with jqwik. Currenlty, this is quite difficult because there is some impedence mismatch:
```Java
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class JqwikAndSpringIntegrationIsDifficult {
    private static TestContextManager testContextManager;

    {
        testContextManager = new TestContextManager(JqwikAndSpringIntegrationIsDifficult.class);
        try {
            testContextManager.beforeTestClass();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
    @Autowired
    EmployeeRepository repository;

    @LocalServerPort
    int randomServerPort;

    @Property
    void test(){}
}
```
The framework offers methods which should be called before/after a class, before/after a test method, before/after a test invocation. However, a single try in jqwik with generated parameters does not necessarily fit any of these methods. In Jupiter the [SpringExtension](https://github.com/spring-projects/spring-framework/blob/master/spring-test/src/main/java/org/springframework/test/context/junit/jupiter/SpringExtension.java) takes care of the integration. If I would transfer it to jqwik, I would create something similar to this code (shamefully stolen):
```Java
public interface SpringSupport implements ResolveParameterHook {

	/**
	 * {@link Store} in which {@code TestContextManagers} are stored,
	 * keyed by test class.
	 */
	 static final Store store = Store.create(SpringExtension.class);


	/**
	 * Delegates to {@link TestContextManager#beforeTestClass}.
	 */
	@BeforeContainer
	public default void beforeAll(PropertyLifecycleContext context) throws Exception {
		getTestContextManager(context).beforeTestClass();
	}

	/**
	 * Delegates to {@link TestContextManager#afterTestClass}.
	 */
	@AfterContainer
	public default void afterAll(PropertyLifecycleContext context) throws Exception {
		try {
			getTestContextManager(context).afterTestClass();
		}
		finally {
			store.remove(context.containerClass());
		}
	}

	/**
	 * Delegates to {@link TestContextManager#prepareTestInstance}.
	 */
	@AfterProperty
	public default void postProcessTestInstance(Object testInstance, PropertyLifecycleContext context) throws Exception {
		getTestContextManager(context).prepareTestInstance(testInstance);
	}

	/**
	 * Delegates to {@link TestContextManager#beforeTestMethod}.
	 */
	@BeforeProperty
	public default void beforeEach(PropertyLifecycleContext context) throws Exception {
		Object testInstance = context.getRequiredTestInstance();
		Method testMethod = context.getRequiredTestMethod();
		getTestContextManager(context).beforeTestMethod(testInstance, testMethod);
	}

	/**
	 * Delegates to {@link TestContextManager#beforeTestExecution}.
	 */
	@BeforeTry
	public default void beforeTestExecution(PropertyLifecycleContext context) throws Exception {
		Object testInstance = context.getRequiredTestInstance();
		Method testMethod = context.getRequiredTestMethod();
		getTestContextManager(context).beforeTestExecution(testInstance, testMethod);
	}

	/**
	 * Delegates to {@link TestContextManager#afterTestExecution}.
	 */
	@AfterTry
	public  default void afterTestExecution(PropertyLifecycleContext context) throws Exception {
		Object testInstance = context.testInstance();
		Method testMethod = context.targetMethod();
		Throwable testException = context.getExecutionException().orElse(null);
		getTestContextManager(context).afterTestExecution(testInstance, testMethod, testException);
	}

	/**
	 * Delegates to {@link TestContextManager#afterTestMethod}.
	 */
	@AfterProperty
	public default void afterEach(PropertyLifecycleContext context) throws Exception {
		Object testInstance = context.getRequiredTestInstance();
		Method testMethod = context.getRequiredTestMethod();
		Throwable testException = context.getExecutionException().orElse(null);
		getTestContextManager(context).afterTestMethod(testInstance, testMethod, testException);
	}

	/**
	 * Determine if the value for the {@link Parameter} in the supplied {@link ParameterContext}
	 * should be autowired from the test's {@link ApplicationContext}.
	 * <p>A parameter is considered to be autowirable if one of the following
	 * conditions is {@code true}.
	 * <ol>
	 * <li>The {@linkplain ParameterContext#getDeclaringExecutable() declaring
	 * executable} is a {@link Constructor} and
	 * {@link TestConstructorUtils#isAutowirableConstructor(Constructor, Class)}
	 * returns {@code true}.</li>
	 * <li>The parameter is of type {@link ApplicationContext} or a sub-type thereof.</li>
	 * <li>{@link ParameterResolutionDelegate#isAutowirable} returns {@code true}.</li>
	 * </ol>
	 * <p><strong>WARNING</strong>: If a test class {@code Constructor} is annotated
	 * with {@code @Autowired} or automatically autowirable (see {@link TestConstructor}),
	 * Spring will assume the responsibility for resolving all parameters in the
	 * constructor. Consequently, no other registered {@link ParameterResolver}
	 * will be able to resolve parameters.
	 * @see #resolveParameter
	 * @see TestConstructorUtils#isAutowirableConstructor(Constructor, Class)
	 * @see ParameterResolutionDelegate#isAutowirable
	 */
	// @Override this mehod is not part of ResolveParameterHook
	public default boolean supportsParameter(ParameterResolutionContext parameterContext, PropertyLifecycleContext propertyContext) {
		Parameter parameter = parameterContext.getParameter();
		Executable executable = parameter.getDeclaringExecutable();
		Class<?> testClass = propertyContext.containerClass();
		return (TestConstructorUtils.isAutowirableConstructor(executable, testClass) ||
				ApplicationContext.class.isAssignableFrom(parameter.getType()) ||
				ParameterResolutionDelegate.isAutowirable(parameter, parameterContext.getIndex()));
	}

	/**
	 * Resolve a value for the {@link Parameter} in the supplied {@link ParameterContext} by
	 * retrieving the corresponding dependency from the test's {@link ApplicationContext}.
	 * <p>Delegates to {@link ParameterResolutionDelegate#resolveDependency}.
	 * @see #supportsParameter
	 * @see ParameterResolutionDelegate#resolveDependency
	 */
	@Override
	@Nullable
	public default Object resolve(ParameterResolutionContext parameterContext, PropertyLifecycleContext propertyContext) {
		Parameter parameter = parameterContext.getParameter();
		int index = parameterContext.getIndex();
		Class<?> testClass = propertyContext.containerClass();
		ApplicationContext applicationContext = getApplicationContext(propertyContext);
		return ParameterResolutionDelegate.resolveDependency(parameter, index, testClass,
				applicationContext.getAutowireCapableBeanFactory());
	}


	/**
	 * Get the {@link ApplicationContext} associated with the supplied {@code PropertyLifecycleContext}.
	 * @param context the current {@code PropertyLifecycleContext} (never {@code null})
	 * @return the application context
	 * @throws IllegalStateException if an error occurs while retrieving the application context
	 * @see org.springframework.test.context.TestContext#getApplicationContext()
	 */
	public static ApplicationContext getApplicationContext(PropertyLifecycleContext context) {
		return getTestContextManager(context).getTestContext().getApplicationContext();
	}

	/**
	 * Get the {@link TestContextManager} associated with the supplied {@code PropertyLifecycleContext}.
	 * @return the {@code TestContextManager} (never {@code null})
	 */
	static TestContextManager getTestContextManager(PropertyLifecycleContext context) {
		Assert.notNull(context, "PropertyLifecycleContext must not be null");
		Class<?> testClass = context.containerClass();
		return store.getOrComputeIfAbsent(testClass, TestContextManager::new, TestContextManager.class); // Also missing from Store
	}

}
``` 

Then I could rewrite my code as follows:

```Java
@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class JqwikAndSpringIntegrationIsDifficult implements SpringSupport {
    @Autowired
    EmployeeRepository repository;

    @LocalServerPort
    int randomServerPort;

    @Property
    void test(){}
}
```

Is this something you intend to support? Is there a way, I could support this effort?

**Comment by jlink** [comment]
@gtrefs https://github.com/jlink/jqwik/issues/41 is about that topic. 
I always had it in the back of my mind when implementing the hooks. My plan is to start working on it as soon as 1.2.4 is done and I’ll keep you in the loop.

**Comment by jlink** [comment]
With https://github.com/jlink/jqwik/commit/89548cb9b6a3035c88fd52918153f8f7fe90a57a `@BeforeContainer` and `@AfterContainer` are supposed to work.

**Comment by jlink** [comment]
With https://github.com/jlink/jqwik/commit/500e248ee692eac0c67a5ba48176d7a879b50810 `@BeforeProperty` and `@AfterProperty` are supposed to work.

**Comment by jlink** [comment]
With https://github.com/jlink/jqwik/commit/b1c078714b0af693f363539994923c91a7a052c2 `@BeforeTry` and `@AfterTry` are supposed to work.

It's now also available in snapshot 1.2.4-SNAPSHOT

**Comment by jlink** [comment]
Documented in userguide:
https://jqwik.net/docs/snapshot/user-guide.html#annotated-lifecycle-methods
